core/src/main/java/org/elasticsearch/common/util/concurrent/SuspendableRefContainer.java

```diff
@@ -0,0 +1,102 @@
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util.concurrent;
+
+import org.elasticsearch.common.lease.Releasable;
+
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Container that represents a resource with reference counting capabilities. Provides operations to suspend acquisition of new references.
+ * This is useful for resource management when resources are intermittently unavailable.
+ *
+ * Assumes less than Integer.MAX_VALUE references are concurrently being held at one point in time.
+ */
+public class SuspendableRefContainer {
+    private static final int TOTAL_PERMITS = Integer.MAX_VALUE;
+    private final Semaphore semaphore;
+
+    public SuspendableRefContainer() {
+        // fair semaphore to ensure that blockAcquisition() does not starve under thread contention
+        this.semaphore = new Semaphore(TOTAL_PERMITS, true);
+    }
+
+    /**
+     * Tries acquiring a reference. Returns reference holder if reference acquisition is not blocked at the time of invocation (see
+     * {@link #blockAcquisition()}). Returns null if reference acquisition is blocked at the time of invocation.
+     *
+     * @return reference holder if reference acquisition is not blocked, null otherwise
+     * @throws InterruptedException if the current thread is interrupted
+     */
+    public Releasable tryAcquire() throws InterruptedException {
+        if (semaphore.tryAcquire(1, 0, TimeUnit.SECONDS)) { // the untimed tryAcquire methods do not honor the fairness setting
+            return () -> semaphore.release();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Acquires a reference. Blocks if reference acquisition is blocked at the time of invocation.
+     *
+     * @return reference holder
+     * @throws InterruptedException if the current thread is interrupted
+     */
+    public Releasable acquire() throws InterruptedException {
+        semaphore.acquire();
+        return () -> semaphore.release();
+    }
+
+    /**
+     * Acquires a reference. Blocks if reference acquisition is blocked at the time of invocation.
+     *
+     * @return reference holder
+     */
+    public Releasable acquireUninterruptibly() {
+        semaphore.acquireUninterruptibly();
+        return () -> semaphore.release();
```

Comment:
this is dangerous, Releasable implements Closeable and that explicitly allows for double closing. I think we have to protecte from double closing on this level!  can we have a simple Releaseable impl that does that?